### PR TITLE
Magnetometer offset cancellation

### PR DESCRIPTION
**Pull request by DusterTheFirst**
Reading through the datasheet of the `lsm303agr` I ran upon section `4.1.2` which details how to implement magnetometer offset cancellation, but also the fact that the magnetometer can do the cancellation automatically when in continuous mode. This feature can be useful for consumers of the library so I attempted to implement it myself.

The implementation, from my observation, looks to be working (the magnetometer data is visually different when the cancellation is on versus off) but I am not very familiar with the layout of this crate and how each piece is used and works together. I tried my best to base my implementation off of the other functions found in this crate, but I cannot guarantee that it matches how the rest of the crate is designed.

### CODE REVIEW

File: src/magnetometer.rs [hunk 1]
```diff
@@ -54,6 +54,25 @@ where
...
         })
     }
+
+    /// Enable or disable the magnetometer's built in offset cancellation
+    ///
+    /// Offset cancellation is **automatically** managed by the device in **continuous** mode.
+    ///
+    /// Offset cancellation has to be **managed by the user** in **single measurement** (OneShot) mode averaging
+    /// two consecutive measurements H<sub>n</sub> and H<sub>n-1</sub>.
+    pub fn mag_offset_cancellation(&mut self, enable: bool) -> Result<(), Error<CommE, PinE>> {
```
Comment (eldruin) [inline src/magnetometer.rs:64]: Could you rather define a method to enable offset calculation and another one to disable it?
Something like `enable_mag_offset_cancellation` and `disable_mag_offset_cancellation`.
(same below)

File: src/magnetometer.rs [hunk 2]
```diff
@@ -54,6 +54,25 @@ where
...
+        let cfg = self.cfg_reg_b_m.bits & 0b11111101;
+
+        let mask = if enable { 1 << 1 } else { 0 };
+
+        self.iface
+            .write_mag_register(Register::CFG_REG_B_M, cfg | mask)?;
+
+        self.cfg_reg_b_m = (cfg | mask).into();
```
Comment (eldruin) [inline src/magnetometer.rs:72]: Could you do these bit flag manipulations by defining variants in the [`BitFlags` `struct`](https://github.com/eldruin/lsm303agr-rs/blob/ba698f7cf1830fc5a6a7d4064d8bb97efdfa729c/src/register_address.rs#L22) and then using the methods `with_high` and `with_low`. See for example [this](https://github.com/eldruin/lsm303agr-rs/blob/3062525017e21730dacd5af219c4576dee004e8f/src/accel_mode_and_odr.rs#L151)
(same below)

**Review by eldruin — COMMENTED** [review]
Nice work, thank you!
A few additional things before merging this:
- Could you add integration tests for the new methods? See for example [here](https://github.com/eldruin/lsm303agr-rs/blob/3062525017e21730dacd5af219c4576dee004e8f/tests/accel_mode_and_odr.rs#L128)
- Could you add the new methods to the method overview in the README and lib.rs?
- Could you add an entry to the Changelog about the new functionality?

**Comment by DusterTheFirst** [comment]
> Could you add integration tests for the new methods?

Looking at other integration tests, I can understand how I would go about setting one up, but I am unsure how I would be able to check that my functions work or what to test exactly. 

My thoughts would be to make sure the correct flags in the registers are set, but there is no way to check that since they are private to the struct. I could just have tests that run the functions and make sure they do not return an Err(_).

If you could clarify what specifically the tests should address that would help me a lot in making them.

**Comment by eldruin** [comment]
What I test is that the correct operations with the correct values are done on the I2C bus, as this should be the ultimate test. For example [here](https://github.com/eldruin/lsm303agr-rs/blob/3062525017e21730dacd5af219c4576dee004e8f/tests/accel_mode_and_odr.rs#L127-L134).
I have not tested for the status of config register variable inside the driver struct explicitly so far. It would be possible to check that this internal status behaves correctly by performing more operations in the test. Then you can see that the bits flipped by previous operations are kept and not simply overwritten.

**Comment by DusterTheFirst** [comment]
Is there any reason for the code duplication between

https://github.com/eldruin/lsm303agr-rs/blob/8bc1f6bae63d2caa98f3312161b05594541482af/src/register_address.rs

and

https://github.com/eldruin/lsm303agr-rs/blob/ba698f7cf1830fc5a6a7d4064d8bb97efdfa729c/tests/common/mod.rs

**Comment by eldruin** [comment]
> Is there any reason for the code duplication between [src/register_address.rs](https://github.com/eldruin/lsm303agr-rs/blob/master/src/register_address.rs) and [tests/common/mod.rs](https://github.com/eldruin/lsm303agr-rs/blob/master/tests/common/mod.rs)

The reason is that the register address is a private module that I would not want to expose and the integration tests (those on top-level under the tests folder) can only access the public interface.
Doing `embedded-hal-mock`-based unit tests instead of integration tests is not possible because `embedded-hal-mock` depends on the `std`.
So far it is just a few constants that will never change so I am fine with the duplication.

Re. running examples on non-linux platforms: what is the reason you want to do that?
Is it because you are developing on a windows/mac? In this case, you can just use [cross](https://github.com/rust-embedded/cross#dependencies) with:
```
cross test --target x86_64-unknown-linux-gnu
```
Works like a charm and the examples can be kept easy.
The linux example is intended for a Raspberry Pi.

**Comment by DusterTheFirst** [comment]
> Works like a charm and the examples can be kept easy.

Good to know. I normally develop on linux but have been using windows against my wishes. I do not have docker installed so I haven't used cross yet, but the tests should all pass in the CI.

**Review by eldruin — APPROVED** [review]
Looks great! Thank you for all your work and happy holidays.

**Pull request merged**